<!DOCTYPE html>
<html lang="en">
<head>
    <title>exceptions.rdoc</title>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width,initial-scale=1">
    
<link rel="stylesheet" href="/css/reset.css" type="text/css" media="screen" data-turbolinks-track="reload" />
<link rel="stylesheet" href="/css/panel.css" type="text/css" media="screen" data-turbolinks-track="reload" />
<link rel="stylesheet" href="/css/main.css" type="text/css" media="screen" data-turbolinks-track="reload" />
<link rel="stylesheet" href="/css/github.css" type="text/css" media="screen" data-turbolinks-track="reload" />
<script src="/js/jquery-3.5.1.min.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/main.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/highlight.pack.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/turbolinks.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/search_index.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/searcher.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/panel/tree.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>
<script src="/js/searchdoc.js" type="text/javascript" charset="utf-8" data-turbolinks-track="reload"></script>

<meta name="data-rel-prefix" content="/">
<meta name="data-tree-keys" content='[]'>

</head>

<body>
    <a class="sr-only sr-only-focusable" href="#content" data-turbolinks="false">Skip to Content</a>
    <a class="sr-only sr-only-focusable" href="#search" data-turbolinks="false">Skip to Search</a>

    <input type="checkbox" id="hamburger" class="panel_checkbox">
<label class="panel_mobile_button" for="hamburger"><span></span> Menu</label>
<nav class="panel panel_tree" id="panel" data-turbolinks-permanent>
  <div class="header">
    <input type="text" placeholder="Search (/) for a class, method, ..." autosave="searchdoc" results="10" id="search" autocomplete="off" tabindex="-1" />
    <label class="panel_mobile_button_close" for="hamburger"><span></span> Close</label>
  </div>
  <div class="tree">
    <ul>
    </ul>
  </div>
  <div class="result">
    <ul>
    </ul>
  </div>
  <a href="links.html" id="links">index</a>
</nav>


    <div class="banner">
        
        <h2>
            exceptions.rdoc
        </h2>
        <ul class="files">
            
            <li>
                ruby/doc/syntax/exceptions.rdoc
                
                    <a href="https://github.com/ruby/ruby/blob/49dc0a12f185d791f1826396035b54cf3fb1771e/doc/syntax/exceptions.rdoc" target="_blank" class="github_url">on GitHub</a>
                
            </li>
            <li>Last modified: 2024-06-28 01:29:37 +0300</li>
        </ul>
    </div>

    <main id="bodyContent">
        <div id="content">
  
    <div class="description">
      
<h1 id="label-Exception+Handling"><a href="../../../../classes/Exception.html"><code>Exception</code></a> Handling</h1>

<p>Exceptions are rescued in a <code>begin</code>/<code>end</code> block:</p>

<pre><code>begin
  # code that might raise
rescue
  # handle exception
end
</code></pre>

<p>If you are inside a method, you do not need to use <code>begin</code> or <code>end</code> unless you wish to limit the scope of rescued exceptions:</p>

<pre><code>def my_method
  # ...
rescue
  # ...
end
</code></pre>

<p>The same is true for a <code>class</code>, <code>module</code>, and <code>block</code>:</p>

<pre><code>[0, 1, 2].map do |i|
  10 / i
rescue ZeroDivisionError
  nil
end
#=&gt; [nil, 10, 5]
</code></pre>

<p>You can assign the exception to a local variable by using <code>=&gt; variable_name</code> at the end of the <code>rescue</code> line:</p>

<pre><code>begin
  # ...
rescue =&gt; exception
  warn exception.message
  raise # re-raise the current exception
end
</code></pre>

<p>By default, <a href="../../../../classes/StandardError.html"><code>StandardError</code></a> and its subclasses are rescued.  You can rescue a specific set of exception classes (and their subclasses) by listing them after <code>rescue</code>:</p>

<pre><code>begin
  # ...
rescue ArgumentError, NameError
  # handle ArgumentError or NameError
end
</code></pre>

<p>You may rescue different types of exceptions in different ways:</p>

<pre><code>begin
  # ...
rescue ArgumentError
  # handle ArgumentError
rescue NameError
  # handle NameError
rescue
  # handle any StandardError
end
</code></pre>

<p>The exception is matched to the rescue section starting at the top, and matches only once.  If an <a href="../../../../classes/ArgumentError.html"><code>ArgumentError</code></a> is raised in the begin section, it will not be handled in the <a href="../../../../classes/StandardError.html"><code>StandardError</code></a> section.</p>

<p>You may retry rescued exceptions:</p>

<pre><code>begin
  # ...
rescue
  # do something that may change the result of the begin block
  retry
end
</code></pre>

<p>Execution will resume at the start of the begin block, so be careful not to create an infinite loop.</p>

<p>Inside a rescue block is the only valid location for <code>retry</code>, all other uses will raise a <a href="../../../../classes/SyntaxError.html"><code>SyntaxError</code></a>.  If you wish to retry a block iteration use <code>redo</code>. See Control Expressions for details.</p>

<p>To always run some code whether an exception was raised or not, use <code>ensure</code>:</p>

<pre><code>begin
  # ...
rescue
  # ...
ensure
  # this always runs
end
</code></pre>

<p>You may also run some code when an exception is not raised:</p>

<pre><code>begin
  # ...
rescue
  # ...
else
  # this runs only when no exception was raised
ensure
  # ...
end
</code></pre>

    </div>
  


  


  
  


  


  

  



  

    

    

    


    


    <!-- Methods -->
    
    
    
  
</div>

    </main>
  </body>
</html>
